### PR TITLE
Created a new proportional plugin option

### PR DESCRIPTION
**Pull request by armstrongmsg**
It uses different parameters for scaling up and down

### CODE REVIEW

File: bigsea-scaler/service/api/controller/plugins/proportional/proportional_alarm.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def __init__(self, actuator, metric_source, trigger_down, trigger_up, min_cap, m
         self.min_cap = min_cap
         self.max_cap = max_cap
         self.metric_rounding = metric_rounding
-        self.heuristic_options = heuristic_options
+        self.heuristic_options_1 = heuristic_options
```
Comment (tellesnobrega) [inline bigsea-scaler/service/api/controller/plugins/proportional/proportional_alarm.py]: This variable name should be something more detailed, this _1 doesn't make much sense

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 2]
```diff
@@ -40,8 +40,13 @@ def setUp(self):
         self.actuator = KVM_Actuator(self.instance_locator, self.remote_kvm)
 
         self.conservative_factor = 1
-        self.heuristic_options = {"heuristic_name":"error_proportional",
+        self.factor_up = 1
+        self.factor_down = 0.5
+        self.heuristic_options_1 = {"heuristic_name":"error_proportional",
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: careful with this heuristic_options names: maybe self.heuristics_options_error_proportional (or self.ho_error_proportional) and the following too.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 3]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        # The method tries to get the metrics correctly
+        self.metric_source.get_most_recent_value.assert_any_call(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_2})
+        
+        # The method tries to get the amount of allocated resources
+        self.actuator.get_allocated_resources.assert_called_once_with(self.instance_name_1)
+        
+        # Remove resources
+        new_cap = self.min_cap
+        # The method tries to adjust the amount of resources
+        self.actuator.adjust_resources.assert_called_once_with({self.instance_name_1:new_cap, self.instance_name_2:new_cap})
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: line too long

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 4]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        
+    def test_alarm_gets_metrics_and_scales_up_error_proportional_up_down(self):
+        #
+        # Case 1: normal scale up
+        #
+        heuristic_options = {"heuristic_name":"error_proportional_up_down",
+                                  "factor_up":self.factor_up, 
+                                  "factor_down":self.factor_down}
+        
+        self.alarm = Proportional_Alarm(self.actuator, self.metric_source, self.trigger_down, self.trigger_up,
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 5]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        self.metric_source.get_most_recent_value = MagicMock()
+        self.metric_source.get_most_recent_value.side_effect = self.metrics
+
+        self.actuator.adjust_resources = MagicMock(return_value=None)
+        self.actuator.get_allocated_resources = MagicMock(return_value=self.allocated_resources)
+
+        self.alarm.check_application_state(self.application_id_0, self.instances)
+
+        # The method tries to get the metrics correctly
+        self.metric_source.get_most_recent_value.assert_any_call(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 6]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+
+        # The method tries to get the amount of allocated resources
+        self.actuator.get_allocated_resources.assert_called_once_with(self.instance_name_1)
+        # Add resources
+        error = self.metrics(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})[1]
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 7]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        self.metric_source.get_most_recent_value = MagicMock()
+        self.metric_source.get_most_recent_value.side_effect = self.metrics
+
+        self.actuator.adjust_resources = MagicMock(return_value=None)
+        self.actuator.get_allocated_resources = MagicMock(return_value=self.allocated_resources)
+
+        self.alarm.check_application_state(self.application_id_0, self.instances)
+
+        # The method tries to get the metrics correctly
+        self.metric_source.get_most_recent_value.assert_any_call(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 8]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        
+        #
+        # Case 2: calculated cap is too high. Use max cap instead
+        #
+        factor_up = 5
+        heuristic_options = {"heuristic_name":"error_proportional_up_down",
+                                  "factor_up":factor_up, 
+                                  "factor_down":self.factor_down}
+        
+        self.alarm = Proportional_Alarm(self.actuator, self.metric_source, self.trigger_down, self.trigger_up,
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 9]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+        # The method tries to get the metrics correctly
+        self.metric_source.get_most_recent_value.assert_any_call(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})
+
+        # The method tries to get the amount of allocated resources
+        self.actuator.get_allocated_resources.assert_called_once_with(self.instance_name_1)
+        # Add resources
+        error = self.metrics(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})[1]
+        new_cap = self.allocated_resources + self.factor_up*abs(error)
+        # The method tries to adjust the amount of resources
+        self.actuator.adjust_resources.assert_called_once_with({self.instance_name_1:new_cap, self.instance_name_2:new_cap})
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py [hunk 10]
```diff
@@ -314,6 +319,131 @@ def test_alarm_metrics_with_different_timestamps(self):
...
+
+        # The method tries to get the metrics correctly
+        self.metric_source.get_most_recent_value.assert_any_call(Proportional_Alarm.ERROR_METRIC_NAME, {"application_id":self.application_id_0})
+
+        # The method tries to get the amount of allocated resources
+        self.actuator.get_allocated_resources.assert_called_once_with(self.instance_name_1)
+        # Add resources
+        new_cap = self.max_cap
+        # The method tries to adjust the amount of resources
+        self.actuator.adjust_resources.assert_called_once_with({self.instance_name_1:new_cap, self.instance_name_2:new_cap})
```
Comment (tellesnobrega) [inline bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: ditto

**Review by tellesnobrega — CHANGES_REQUESTED** [review]

File: bigsea-scaler/service/api/controller/plugins/proportional/proportional_alarm.py (hunk 1 above)
Comment (armstrongmsg) [reply bigsea-scaler/service/api/controller/plugins/proportional/proportional_alarm.py]: OK (Eclipse refactoring tool...)

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 2 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 3 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 4 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 5 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 6 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 9 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 8 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 7 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

File: bigsea-scaler/test/service/api/controller/test_proportional_alarm.py (hunk 10 above)
Comment (armstrongmsg) [reply bigsea-scaler/test/service/api/controller/test_proportional_alarm.py]: Fixed.

**Pull request merged**